x/mongo/driver/ocsp/cache.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package ocsp
+
+import (
+	"crypto"
+	"sync"
+	"time"
+
+	"golang.org/x/crypto/ocsp"
+)
+
+type cacheKey struct {
+	HashAlgorithm  crypto.Hash
+	IssuerNameHash string
+	IssuerKeyHash  string
+	SerialNumber   string
+}
+
+// Cache represents an OCSP cache.
+type Cache interface {
+	Put(*ocsp.Request, *Response) *Response
+	Get(request *ocsp.Request) *Response
+}
+
+// ConcurrentCache is an implementation of ocsp.Cache that's safe for concurrent use.
+type ConcurrentCache struct {
+	cache map[cacheKey]*Response
+	sync.Mutex
+}
+
+// NewCache creates an empty OCSP cache.
+func NewCache() *ConcurrentCache {
+	return &ConcurrentCache{
+		cache: make(map[cacheKey]*Response),
+	}
+}
+
+// Put stores the given request into the cache. If the request is already in the cache, the existing entry will be
+// replaced.
+func (c *ConcurrentCache) Put(request *ocsp.Request, response *Response) *Response {
+	unknown := response.Status == ocsp.Unknown
```

Comment:
There is currently not such an invariant because as I mentioned in a comment to David, the driver's handling of panics in background routines is not great, but this is something I can see doing for follow-on work.